Security: CVE-2025-69534 – Remote DoS via malformed <![ sequence in Python-Markdown 3.8

## Description
While fuzz-testing the Python-Markdown library (version 3.8) with the `extra` extension enabled, I found that certain malformed inputs containing `<![` sequences cause the parser to throw uncaught exceptions and crash. These inputs appear to break the parser's handling of XML-style marked sections, leading to errors like "expected name token" or "unknown status keyword."

## Steps to Reproduce
1. Run the following test script (using Python-Markdown 3.8 with `extra` extension):

```python
import markdown

print(markdown.__version__)

crash_inputs = [
    "<![",
    "<![>og))/uw_     f{tv+pAr$Ss+[6;^{=<:>g2oV|.pdTMu(Q-E#",
    "<![ g'\"7z5r7cojSO;2LAo0(1Vv5G>,-P",
]

for i, crash_input in enumerate(crash_inputs, 1):
    print(f"\nTesting crash input #{i}:\n{repr(crash_input)}")
    try:
        output = markdown.markdown(crash_input, extensions=["extra"])
        print("No crash, output:")
        print(output)
    except Exception as e:
        print(f"Crash confirmed! Exception:\n{e}")

```

2. Observe that the first three inputs cause exceptions, crashing the parser.

## Expected Behavior
The parser should gracefully handle or sanitize malformed `<![` marked sections without raising uncaught exceptions.

## Actual Behavior
Uncaught exceptions are raised, such as:

```python
3.8

Testing crash input #1:
'<!['
Crash confirmed! Exception:
expected name token at '<![\n\n'

Testing crash input #2:
'<![>og))/uw_     f{tv+pAr$Ss+[6;^{=<:>g2oV|.pdTMu(Q-E#'
Crash confirmed! Exception:
expected name token at '<![>og))/uw_     f{t'

Testing crash input #3:
'<![ g\'"7z5r7cojSO;2LAo0(1Vv5G>,-P'
Crash confirmed! Exception:
expected name token at '<![ g\'"7z5r7cojSO;2L'
```


- `expected name token at '<![\n\n'`
- `expected name token at '<![>og))/uw_ f{t'`
- `expected name token at '<![ g\'"7z5r7cojSO;2L'`



## Impact
- Potential Denial of Service by crashing apps parsing untrusted Markdown.
- Stability risk in applications relying on Python-Markdown.
- Possible information leakage via stack traces if errors are exposed.

## Environment
- Python-Markdown version: 3.8
- Python version: 3.11
- OS: Linux and Windows 11

## Additional Notes
I performed fuzz testing and isolated these inputs causing crashes. I'm happy to provide more test cases or logs if needed.

Thank you for your attention to this issue!

I can confirm the behavior. Thanks for the report. However, there is no need for `extra` to get the reported behavior. 

``` python
>>> markdown.markdown('<![')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\code\md\markdown\core.py", line 482, in markdown
    return md.convert(text)
           ^^^^^^^^^^^^^^^^
  File "C:\code\md\markdown\core.py", line 354, in convert
    self.lines = prep.run(self.lines)
                 ^^^^^^^^^^^^^^^^^^^^
  File "C:\code\md\markdown\preprocessors.py", line 89, in run
    parser.feed(source)
  File "C:\Users\wlimberg\AppData\Roaming\uv\python\cpython-3.12.9-windows-x86_64-none\Lib\html\parser.py", line 111, in feed
    self.goahead(0)
  File "C:\Users\wlimberg\AppData\Roaming\uv\python\cpython-3.12.9-windows-x86_64-none\Lib\html\parser.py", line 179, in goahead
    k = self.parse_html_declaration(i)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\code\md\markdown\htmlparser.py", line 278, in parse_html_declaration
    return super().parse_html_declaration(i)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wlimberg\AppData\Roaming\uv\python\cpython-3.12.9-windows-x86_64-none\Lib\html\parser.py", line 264, in parse_html_declaration
    return self.parse_marked_section(i)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wlimberg\AppData\Roaming\uv\python\cpython-3.12.9-windows-x86_64-none\Lib\_markupbase.py", line 144, in parse_marked_section
    sectName, j = self._scan_name( i+3, i )
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wlimberg\AppData\Roaming\uv\python\cpython-3.12.9-windows-x86_64-none\Lib\_markupbase.py", line 390, in _scan_name
    raise AssertionError(
AssertionError: expected name token at '<![\n\n'
```

Looks like we need to catch the `AssertionError` in `parse_html_declaration` and handle it somehow.

This is a high priority issue as crashes from Markdown input are unacceptable.

--- a/docs/changelog.md
+++ b/docs/changelog.md
@@ -13,6 +13,7 @@ See the [Contributing Guide](contributing.md) for details.
 
 ### Fixed
 
+* Ensure incomplete markup declaration in raw HTML doesn't crash parser (#1534).
 * Fixed dropped content in `md_in_html` (#1526).
 * Fixed HTML handling corner case that prevented some content from not being rendered (#1528).
 

--- a/markdown/extensions/md_in_html.py
+++ b/markdown/extensions/md_in_html.py
@@ -281,7 +281,7 @@ class HTMLExtractorExtra(HTMLExtractor):
     def parse_html_declaration(self, i: int) -> int:
         if self.at_line_start() or self.intail or self.mdstack:
             if self.rawdata[i:i+3] == '<![' and not self.rawdata[i:i+9] == '<![CDATA[':
-                # We have encountered the bug in #1534 (Python bug gh-77057).
+                # We have encountered the bug in #1534 (Python bug `gh-77057`).
                 # Provide an override until we drop support for Python < 3.13.
                 return self.parse_bogus_comment(i)
             # The same override exists in `HTMLExtractor` without the check

--- a/markdown/extensions/md_in_html.py
+++ b/markdown/extensions/md_in_html.py
@@ -280,6 +280,10 @@ class HTMLExtractorExtra(HTMLExtractor):
 
     def parse_html_declaration(self, i: int) -> int:
         if self.at_line_start() or self.intail or self.mdstack:
+            if self.rawdata[i:i+3] == '<![' and not self.rawdata[i:i+9] == '<![CDATA[':
+                # We have encountered the bug in #1534 (Python bug gh-77057).
+                # Provide an override until we drop support for Python < 3.13.
+                return self.parse_bogus_comment(i)
             # The same override exists in `HTMLExtractor` without the check
             # for `mdstack`. Therefore, use parent of `HTMLExtractor` instead.
             return super(HTMLExtractor, self).parse_html_declaration(i)

--- a/markdown/htmlparser.py
+++ b/markdown/htmlparser.py
@@ -276,7 +276,7 @@ class HTMLExtractor(htmlparser.HTMLParser):
     def parse_html_declaration(self, i: int) -> int:
         if self.at_line_start() or self.intail:
             if self.rawdata[i:i+3] == '<![' and not self.rawdata[i:i+9] == '<![CDATA[':
-                # We have encountered the bug in #1534 (Python bug gh-77057).
+                # We have encountered the bug in #1534 (Python bug `gh-77057`).
                 # Provide an override until we drop support for Python < 3.13.
                 return self.parse_bogus_comment(i)
             return super().parse_html_declaration(i)

--- a/markdown/htmlparser.py
+++ b/markdown/htmlparser.py
@@ -275,6 +275,10 @@ class HTMLExtractor(htmlparser.HTMLParser):
 
     def parse_html_declaration(self, i: int) -> int:
         if self.at_line_start() or self.intail:
+            if self.rawdata[i:i+3] == '<![' and not self.rawdata[i:i+9] == '<![CDATA[':
+                # We have encountered the bug in #1534 (Python bug gh-77057).
+                # Provide an override until we drop support for Python < 3.13.
+                return self.parse_bogus_comment(i)
             return super().parse_html_declaration(i)
         # This is not the beginning of a raw block so treat as plain data
         # and avoid consuming any tags which may follow (see #1066).

--- a/tests/test_syntax/blocks/test_html_blocks.py
+++ b/tests/test_syntax/blocks/test_html_blocks.py
@@ -1275,6 +1275,13 @@ class TestHTMLBlocks(TestCase):
             )
         )
 
+    def test_not_actually_cdata(self):
+        # Ensure bug reported in #1534 is avoided.
+        self.assertMarkdownRenders(
+            '<![',
+            '<p>&lt;![</p>'
+        )
+
     def test_raw_cdata_code_span(self):
         self.assertMarkdownRenders(
             self.dedent(

